### PR TITLE
Backport bit manipulation facility from C++20 standard library header `<bit>`

### LINKED ISSUE #3828: Deprecate/move Kokkos::log2(unsigned) -> int
`int log2(unsigned)` is currently defined in `Kokkos::`.  This was an oversight, it should have been in `Impl::` and this will become an issue if we decide to promote the mathematical functions that were recently added to `Experimental::` because it would hide the templated overload `double log2(IntegralType)` when passing an `unsigned` argument.
https://github.com/kokkos/kokkos/blob/5c32ef063f44b3be81b21ff2aacde177dd02aa21/core/src/impl/Kokkos_BitOps.hpp#L58-L77

We need to formulate a plan to deprecate it, and later remove it.
Note, that C++20 adds to the standard library a facility that provides this functionality in the [`<bit>` header](https://eel.is/c++draft/bit).

### PR DESCRIPTION
**Pull request by dalg24**
~~Fix #3828~~

**See documentation in kokkos/kokkos-core-wiki#300**

* ~~Deprecate `Kokkos::log2(unsigned) -> int`~~ done in #4595
* ~~Provide alternative `Impl::bit_log2(UnsignedInteger) -> UnsignedInteger` for internal use only~~ withdrawn (might come back with it in a follow up PR)
* Backport bit manipulation facility from C++20 https://eel.is/c++draft/bit
* ~~Remove `Impl::bit_count(unsigned int)` in favor of `Experimental::popcount(UnsignedInteger) -> UnsignedInteger`~~ withdrawn (will propose in a follow up PR)

#### NOTES
* I have already implemented `[bit.rot]` and I intend to remove `Impl::rotate_right(unsigned, int) -> unsigned` in the `Bitset` implementation but I was planing to do that in a follow-up PR
* I intend to incrementally replace everything in `<impl/Kokkos_BitOps.hpp>` until we can get rid of the header
* I will also attempt to replace the `integral_power_of_two*` facility in `impl/Kokkos_Traits.hpp` but I would rather deal with it later
* I only used GCC built-in functions on the host-side for `Experimental::*_builtin` function templates.  We can specialize further and add Intel intrinsics and such later.

### CODE REVIEW

**Review by ajpowelsnl — APPROVED** [review]
The PR generally looks good, and there's no reason to think it shouldn't be merged.  But I do have a basic question (perhaps I should know the answer, but I don't), why does Kokkos need to manipulate at the bit level?  To deal with various proclivities and folkways of different compilers and backends?

**Comment by PhilMiller** [comment]
This is part of a broader theme of Kokkos broadly mimicking C++ standard library functionality with `__host__ __device__` annotations (or their moral equivalent) so that applications can use them on GPUs. See also `Kokkos::pair`, `Kokkos::complex`, `Kokkos::array`, math routines (`pow`, `sqrt`, etc), and much more.

File: core/unit_test/TestBitManipulationBuiltins.hpp [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(has_single_bit);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(bit_ceil);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(bit_floor);
+DEFINE_BIT_MANIPULATION_FUNCTION_EVAL(bit_width);
+
+template <class Space, class Func, class Arg, std::size_t N>
+struct TestBitManipFunction {
+  Arg val_[N];
+  TestBitManipFunction(const Arg (&val)[N]) {
+    std::copy(val, val + N, val_);
```
Comment (masterleinad) [inline core/unit_test/TestBitManipulationBuiltins.hpp:61]: Maybe also make sure here that the `*_builtin` can be used in a `constexpr` context?

File: core/src/Kokkos_BitManipulation.hpp [hunk 2]
```diff
@@ -0,0 +1,356 @@
...
+
+template <class T>
+inline constexpr bool is_standard_unsigned_integer_type_v =
+    std::is_same_v<T, unsigned char> || std::is_same_v<T, unsigned short> ||
+    std::is_same_v<T, unsigned int> || std::is_same_v<T, unsigned long> ||
+    std::is_same_v<T, unsigned long long>;
+
+}  // namespace Kokkos::Impl
+
+namespace Kokkos {
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Kokkos_BitManipulation.hpp:67]: No I want to provide them in `Kokkos::`.  It is a backport from a C++20 facility.  There is no unknown on what we want it to look like.

File: core/unit_test/TestBitManipulationBuiltins.hpp (hunk 1 above)
Comment (dalg24) [reply core/unit_test/TestBitManipulationBuiltins.hpp:61]: ??  The `Experimental::*_builtin` function templates are not declared `constexpr`.

File: core/src/Kokkos_BitManipulation.hpp [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+  using ::Kokkos::Experimental::digits_v;
+  return digits_v<T> - countl_zero(x);
+}
+//</editor-fold>
+
+}  // namespace Kokkos
+
+namespace Kokkos::Impl {
+
+#if defined(KOKKOS_COMPILER_CLANG) || defined(KOKKOS_COMPILER_GCC)
```
[the remark this answers is not in the document]
Comment (dalg24) [reply core/src/Kokkos_BitManipulation.hpp:160]: I'd prefer fiddling with more compiler and other intrinsics later.

File: core/unit_test/TestBitManipulationBuiltins.hpp (hunk 1 above)
Comment (nliber) [reply core/unit_test/TestBitManipulationBuiltins.hpp:61]: The public functions are callable in a `constexpr` context (TestBitManipulation.cpp specifically tests that).

Given that, what is the use case behind `*_builtin` being directly useable in a `constexpr` context?

**Review by nliber — APPROVED** [review]

File: core/unit_test/TestBitManipulationBuiltins.hpp (hunk 1 above)
Comment (masterleinad) [reply core/unit_test/TestBitManipulationBuiltins.hpp:61]: I missed the test.

**Comment by masterleinad** [comment]
> The PR generally looks good, and there's no reason to think it shouldn't be merged. But I do have a basic question (perhaps I should know the answer, but I don't), why does Kokkos need to manipulate at the bit level? To deal with various proclivities and folkways of different compilers and backends?

We already have a much of places where we would use these functions in the library.

**Review by masterleinad — APPROVED** [review]
Looks OK to me.

**Comment by dalg24** [comment]
(I figured out the failure with NVCC.  It was a few stray constexpr qualifiers on builtin device side function templates.)

**Comment by dalg24** [comment]
Retest this please

**Comment by dalg24** [comment]
Retest this please

File: core/unit_test/TestBitManipulationBuiltins.hpp [hunk 4]
```diff
@@ -0,0 +1,435 @@
...
+      UInt(127),
+      UInt(128),
+      UInt(129),
+      UInt(130),
+      UInt(max),
+  });
+}
+
+TEST(TEST_CATEGORY, bit_manip_countr_zero) {
+  test_bit_manip_countr_zero<unsigned char>();
```
Comment (dalg24) [inline core/unit_test/TestBitManipulationBuiltins.hpp:207]: There is an issue with SYCL
```4: [ RUN      ] sycl.bit_manip_countr_zero
4: value at 1 which is 248 was expected to be 0
4: value at 2 which is 249 was expected to be 1
4: value at 3 which is 248 was expected to be 0
4: value at 4 which is 250 was expected to be 2
4: value at 5 which is 248 was expected to be 0
4: value at 6 which is 249 was expected to be 1
4: value at 7 which is 248 was expected to be 0
4: value at 8 which is 251 was expected to be 3
4: value at 9 which is 248 was expected to be 0
4: value at 7e which is 249 was expected to be 1
4: value at 7f which is 248 was expected to be 0
4: value at 80 which is 255 was expected to be 7
4: value at 81 which is 248 was expected to be 0
4: value at 82 which is 249 was expected to be 1
4: value at 7f which is 248 was expected to be 0
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:67: Failure
4: Expected equality of these values:
4:   errors
4:     Which is: 15
4:   0
4: Failed check no error for countr_zero(unsigned char)
4: [  FAILED  ] sycl.bit_manip_countr_zero (9 ms)

File: core/unit_test/TestBitManipulationBuiltins.hpp [hunk 5]
```diff
@@ -0,0 +1,435 @@
...
+      UInt(126),
+      UInt(127),
+      UInt(128),
+      UInt(max - 1),
+      UInt(max),
+  });
+}
+
+TEST(TEST_CATEGORY, bit_manip_countr_one) {
+  test_bit_manip_countr_one<unsigned char>();
```
Comment (dalg24) [inline core/unit_test/TestBitManipulationBuiltins.hpp:243]:
```
4: [ RUN      ] sycl.bit_manip_countr_one
4: value at 0 which is 248 was expected to be 0
4: value at 1 which is 249 was expected to be 1
4: value at 2 which is 248 was expected to be 0
4: value at 3 which is 250 was expected to be 2
4: value at 4 which is 248 was expected to be 0
4: value at 5 which is 249 was expected to be 1
4: value at 6 which is 248 was expected to be 0
4: value at 7 which is 251 was expected to be 3
4: value at 8 which is 248 was expected to be 0
4: value at 9 which is 249 was expected to be 1
4: value at 7e which is 248 was expected to be 0
4: value at 7f which is 255 was expected to be 7
4: value at 80 which is 248 was expected to be 0
4: value at 7e which is 248 was expected to be 0
4: value at 7f which is 255 was expected to be 7
4: /var/jenkins/workspace/Kokkos/core/unit_test/TestBitManipulationBuiltins.hpp:67: Failure
4: Expected equality of these values:
4:   errors
4:     Which is: 15
4:   0
4: Failed check no error for countr_one(unsigned char)

File: core/unit_test/TestBitManipulationBuiltins.hpp (hunk 4 above)
Comment (dalg24) [reply core/unit_test/TestBitManipulationBuiltins.hpp]: Daniel reported it works on Intel GPUs.  I disabled on NVIDIA.

**Comment by dalg24** [comment]
SYCL build passed with the last commit.  HIP failure is unrelated.

File: core/unit_test/TestBitManipulationBuiltins.hpp [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+  do_test_bit_manip_function<TEST_EXECSPACE, BitManipFunction_##FUNC>
+
+template <class UInt>
+void test_bit_manip_countl_zero() {
+  using Kokkos::Experimental::countl_zero_builtin;
+  static_assert(noexcept(countl_zero_builtin(UInt())));
+  static_assert(std::is_same_v<decltype(countl_zero_builtin(UInt())), int>);
+  constexpr auto max =
+      Kokkos::Experimental::finite_max_v<std::make_signed_t<UInt>>;
+  //                                     ^^^^^^^^^^^^^^^^^^
```
Comment (dalg24) [inline core/unit_test/TestBitManipulationBuiltins.hpp:96]: Didn't mean to keep that change...

**Review by nmm0 — APPROVED** [review]
Looks good to me, we just need to file an issue for the eventual merging of these paths

**Comment by nmm0** [comment]
I promised to myself that when I would review I would ask for docs/changelog entry. You already have the docs, so I guess we should add to the changelog of whatever this is targeting.

**Comment by dalg24** [comment]
Ignoring HIP failure

**Pull request merged**